source/manual/architecture-deep-dive.html.md

```diff
@@ -0,0 +1,488 @@
+[govuk-crawler-worker]: https://github.com/alphagov/govuk_crawler_worker
+
+#### Routing on the CDN
+
+As well as for caching, Varnish is used for the redirection from `gov.uk` to
+`www.gov.uk`, which is configured in Varnish Configuration Language (VCL) and
+uploaded directly to Fastly via [govuk-cdn-config].
+
+Other redirects that happen at the Fastly level include [bouncer]: a GOV.UK
+application responsible for redirecting traffic from old pre-GOV.UK websites.
+This is configured via [transition] and [transition-config]. Read
+[Transition architecture] for more detail.
+
+[bouncer]: https://github.com/alphagov/bouncer
+[govuk-cdn-config]: https://github.com/alphagov/govuk-cdn-config
+[Transition architecture]: /manual/transition-architecture.html
+[transition]: https://github.com/alphagov/transition
+[transition-config]: https://github.com/alphagov/transition-config
+
+### Routing on GOV.UK
+
+#### Getting to the 'router' application
+
+![](images/load-balancer.png)
+
+Some requests make it through the CDN and cache layers to 'origin'. Origin is
+a stack of computers in the cloud - in this case, AWS - and its entry point is
+a load balancer.
+
+The load balancer knows based on the hostname which machine 'class' to route
+to. Different classes of machine run different sets of GOV.UK applications.
+How many machines are allocated to a class - and how big those machines are -
+is configured using [Terraform], in [govuk-aws]. What runs on each machine
+class is configured in govuk-puppet, a file and process management system we'll
+cover in more detail later.
+
+![](images/nginx-routing.png)
+
+External requests are routed to a 'cache' machine, where the request is
+received by an [Nginx] web server running on the machine. Nginx uses
+Varnish for its caching: if Varnish has the route response in its cache,
+that is returned, otherwise Nginx handles the route. Nginx is configured
+with govuk-puppet.
+
+Nginx proxies most requests to [router], which is a GOV.UK maintained
+application running on the cache machines. However, Nginx does proxy some
+routes directly to other apps, such as [asset URLs] being
+[routed to asset-manager][asset-proxy].
+
+[asset URLs]: https://github.com/alphagov/govuk-puppet/blob/881cbcdef477332948e92b88ecd04a830fd02337/hieradata/common.yaml#L1176-L1178
+[asset-proxy]: https://github.com/alphagov/govuk-puppet/blob/a114dd5f80d789be368573545dc56fe388c0ac58/modules/router/templates/assets_origin.conf.erb#L53-L67
+[govuk-aws]: https://github.com/alphagov/govuk-aws
+[Nginx]: https://www.nginx.com/
+[router]: https://github.com/alphagov/router
+[Terraform]: https://www.terraform.io/
+
+#### Routing via 'router'
+
+'Router' is a reverse proxy app written in Go. It is designed to be fast,
+storing all known routes in memory using a [prefix trie], which it loads
+from [Redis] - an in-memory database - at startup. Redis is populated by
+reading from a MongoDB database of all known routes.
```

Comment:
I'm pretty sure it's just MongoDB that's involved. The Router reads the data from MongoDB, in to memory.

Reply:
Continuing the conversation in https://github.com/alphagov/govuk-developer-docs/pull/1811#discussion_r407050257.